### PR TITLE
chore: use new notification to start tests, and also to inform users …

### PR DESCRIPTION
**Pull request by ajbt200128**
…about scan workspace

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by ajbt200128** [comment]
tests won't pass until https://github.com/semgrep/semgrep/pull/10145 is released

File: src/env.ts [hunk 1]
```diff
@@ -89,6 +98,11 @@ export class Environment {
     return this._client;
   }
 
+  emitStartupEvent(): void {
+    console.log("FIRE!");
```
[the remark this answers is not in the document]
Comment (brandonspark) [reply src/env.ts:102]: probably want this removed

File: src/commands.ts [hunk 2]
```diff
@@ -97,11 +97,33 @@ export function registerCommands(env: Environment): void {
...
 
   vscode.commands.registerCommand("semgrep.scanWorkspace", async () => {
-    await env.client?.sendNotification(scanWorkspace, { full: false });
+    const onlyGitDirty = env.config.onlyGitDirty;
+    if (!onlyGitDirty) {
+      vscode.window.showInformationMessage(
+        'Semgrep is now only scanning files and lines that have been changed since the last commit. You can disable this in settings by unchecking "Only Git Dirty", or by running "Scan all files in workspace"'
+      );
+      env.config.onlyGitDirty = true;
+      env.onRulesRefreshed(() =>
```
Comment (brandonspark) [inline src/commands.ts:106]: looks like this code is resting on an implicit assumption that sending `scanWorkspace` will cause `refreshRules`, which will cause an `onRulesRefreshed`. this is all fair, but i wish it were a bit more telegraphed. i think even just a small comment here commenting on that would be fine

File: src/lsp.ts [hunk 3]
```diff
@@ -248,34 +250,15 @@ async function start(env: Environment): Promise<void> {
...
-        if (params?.value?.kind == "end") {
-          env.logger.log("Rules loaded");
-          resolve();
-        }
-      });
-    }
-  });
+  const notificationHandler: NotificationHandler0 = () => {
+    env.logger.log("Rules loaded");
+    env.emitStartupEvent();
```
Comment (brandonspark) [inline src/lsp.ts:256]: isn't this going to fire on _every_ rules refresh from now on, including those which do not happen at startup? should `emitStartupEvent` be traded for a different name?

File: src/env.ts [hunk 4]
```diff
@@ -41,7 +49,8 @@ export class Environment {
     readonly channel: OutputChannel,
     readonly logger: Logger,
     public version: string = "",
-    public startupPromise?: Promise<void>
+    private _rulesRefreshedEmitter: EventEmitter<void> = new EventEmitter<void>(),
```
Comment (brandonspark) [inline src/env.ts:52]: a small comment on `EventEmitter` and how this `onRulesRefreshed` is to be used would be nice. basically, i just had to read this to understand that its a way for us to get `emitStartupEvent` to access the command code

**Review by brandonspark — APPROVED** [review]
a lot more descriptive than the engine-side notifications!

File: src/commands.ts (hunk 2 above)
Comment (ajbt200128) [reply src/commands.ts:106]: yea, that's fair that it's non obvious. I'll add a comment explaining it

File: src/lsp.ts (hunk 3 above)
Comment (ajbt200128) [reply src/lsp.ts:256]: Oh that's true, will rename. It also means that we'll do a scan workspace everytime afterwards, so I'll fix that

**Pull request merged**